crates/bitwarden-crypto/src/safe/password_protected_key_envelope.rs

```diff
@@ -0,0 +1,572 @@
+        let mut cose_encrypt = coset::CoseEncryptBuilder::new()
+            .add_recipient({
+                let mut recipient = coset::CoseRecipientBuilder::new()
+                    .unprotected(kdf_settings.into())
+                    .build();
+                recipient.protected.header.alg = Some(coset::Algorithm::PrivateUse(ALG_ARGON2ID13));
+                recipient
+            })
+            .protected(HeaderBuilder::from(content_format).build())
+            .create_ciphertext(&key_to_seal_bytes, &[], |data, aad| {
+                let ciphertext = xchacha20::encrypt_xchacha20_poly1305(&envelope_key, data, aad);
+                nonce.copy_from_slice(&ciphertext.nonce());
+                ciphertext.encrypted_bytes().to_vec()
+            })
+            .build();
+        cose_encrypt.unprotected.iv = nonce.into();
+
+        Ok(PasswordProtectedKeyEnvelope {
+            _phantom: PhantomData,
+            cose_encrypt,
+        })
+    }
+
+    /// Unseals a symmetric key from the password-protected envelope, and stores it in the key store
+    /// context.
+    pub fn unseal(
+        &self,
+        target_keyslot: Ids::Symmetric,
+        password: &str,
+        ctx: &mut KeyStoreContext<Ids>,
+    ) -> Result<Ids::Symmetric, PasswordProtectedKeyEnvelopeError> {
+        let key = self.unseal_ref(password)?;
+        #[allow(deprecated)]
+        ctx.set_symmetric_key(target_keyslot, key)
+            .map_err(|_| PasswordProtectedKeyEnvelopeError::KeyStoreError)?;
+        Ok(target_keyslot)
+    }
+
+    fn unseal_ref(
+        &self,
+        password: &str,
+    ) -> Result<SymmetricCryptoKey, PasswordProtectedKeyEnvelopeError> {
+        // There must be exactly one recipient in the COSE Encrypt object, which contains the KDF
+        // parameters.
+        if self.cose_encrypt.recipients.len() != 1 {
+            return Err(PasswordProtectedKeyEnvelopeError::ParsingError(
+                "Invalid number of recipients".to_string(),
+            ));
+        }
+
+        let recipient = self.cose_encrypt.recipients.first().ok_or(
+            PasswordProtectedKeyEnvelopeError::ParsingError("Missing recipient".to_string()),
+        )?;
+        if recipient.protected.header.alg != Some(coset::Algorithm::PrivateUse(ALG_ARGON2ID13)) {
+            return Err(PasswordProtectedKeyEnvelopeError::ParsingError(
+                "Unknown or unsupported KDF algorithm".to_string(),
+            ));
+        }
+
+        let kdf_settings: Argon2RawSettings =
+            (&recipient.unprotected).try_into().map_err(|_| {
+                PasswordProtectedKeyEnvelopeError::ParsingError(
+                    "Invalid or missing KDF parameters".to_string(),
+                )
+            })?;
+        let envelope_key = derive_key(&kdf_settings, password)
+            .map_err(|_| PasswordProtectedKeyEnvelopeError::KdfError)?;
+        let nonce: [u8; 24] = self
```

Comment:
```suggestion
        let nonce: [u8; crate::xchacha20::NONCE_SIZE] = self
```

Should we be using the same constant as we did in `seal_ref_with_settings`?